packages/logs/test/logs.spec.ts

```diff
@@ -0,0 +1,227 @@
+import { mockModule, unmockModules } from '../../../test/unit/mockModule'
+
+import {
+  Configuration,
+  Context,
+  DEFAULT_CONFIGURATION,
+  ErrorMessage,
+  ErrorObservable,
+  ErrorOrigin,
+  noop,
+  Observable,
+} from '@datadog/browser-core'
+import sinon from 'sinon'
+
+import { Logger, LogsMessage, StatusType } from '../src/logger'
+import { startLogs } from '../src/logs'
+
+interface SentMessage extends LogsMessage {
+  logger?: { name: string }
+  view: {
+    id?: string
+    referrer?: string
+    url: string
+  }
+}
+
+function getLoggedMessage(server: sinon.SinonFakeServer, index: number) {
+  return JSON.parse(server.requests[index].requestBody) as SentMessage
+}
+const DEFAULT_INIT_CONFIGURATION = { clientToken: 'xxx' }
+const FAKE_DATE = 123456
+const SESSION_ID = 'session-id'
+const configuration: Partial<Configuration> = {
+  ...DEFAULT_CONFIGURATION,
+  logsEndpoint: 'https://localhost/v1/input/log',
+  maxBatchSize: 1,
+  service: 'Service',
+}
+
+interface Rum {
+  getInternalContext(startTime?: number): any | undefined
+}
+declare global {
+  interface Window {
+    DD_RUM?: Rum
+  }
+}
+
+describe('logs', () => {
+  let sessionIsTracked: boolean
+  let configurationOverrides: Partial<Configuration>
+  let server: sinon.SinonFakeServer
+  let errorObservable: ErrorObservable
+
+  beforeEach(() => {
+    sessionIsTracked = true
+    mockModule('./packages/logs/src/loggerSession.ts', () => ({
+      startLoggerSession() {
+        return {
+          getId: () => (sessionIsTracked ? SESSION_ID : undefined),
+          isTracked: () => sessionIsTracked,
+        }
+      },
+    }))
+
+    configurationOverrides = {}
+    errorObservable = new Observable<ErrorMessage>()
+    mockModule('./packages/core/src/init.ts', () => ({
+      commonInit() {
+        return {
+          errorObservable,
+          configuration: { ...(configuration as Configuration), ...configurationOverrides },
+          internalMonitoring: { setExternalContextProvider: () => undefined },
+        }
+      },
+    }))
```

Comment:
It seems to me that you are trying to change the behavior of inner components of your test subject.

What about updating the design in order that your test subject take these components as dependencies?
it would avoid to introduce the `mockModule` technique and would be more consistent with what we are doing elsewhere in the code base.

Reply:
I introduced module mocking *because* you didn't like injecting initialisation values for tests: https://github.com/DataDog/browser-sdk/pull/545#discussion_r494974626
> it handles the init logic, but take the "base" init as a dependency, which make things harder to follow and I don't see the need for it.

Anyway, I really like mocking modules. This is the way `jest` is working. It allows to test the code without having to inject everything. I think it is a superior design to the RUM specBuilder we have in the RUM side.

If you find it confusing I'll revert the `baseInit` function.

Reply:
My previous comment was probably not clear enough, in the previous approach:
```ts
export function makeLogsGlobal(
  baseInit: (
    configuration: LogsUserConfiguration
  ) => {
    session: LoggerSession
    configuration: Configuration
    errorObservable: ErrorObservable
    internalMonitoring: InternalMonitoring
  }
)
```
A Part of the init was made on this base function and another part in the makeLogsGlobal which was hard to follow.
Base init function returned 4 objects which could be hidden from makeLogsGlobal, providing a better encapsulation of the init logic from the evaluation logic.

Now, with something like `makeLogsGlobal(startLog)`, by stubbing `startLog`, we could easily unit test the pre-init logic, the conf check logic and the sendLog strategy switch without needing to be aware of session, errorObservable, ...

> I think it is a superior design to the RUM specBuilder we have in the RUM side.

From my point of view, it is not what we should compare. 
In the current code base, we inject dependencies as parameters. It allows us to easily stub these dependencies in unit test.
With the `mockModule` technique, I can see several drawbacks:
- unit tests are now coupled with webpack and its module cache management
- unit tests are now coupled to some file path, ex:`mockModule('./packages/logs/src/logs.ts', () =>` 
- no type check are performed on mocked method output which could easily lead to confusing errors

Reply:
I reverted the module mocks and brought back dependency injection.